src/dataframe/dataframe.jl

```diff
@@ -1317,3 +1317,79 @@ end
 function permutecols!(df::DataFrame, p::AbstractVector{Symbol})
     permutecols!(df, index(df)[p])
 end
+
+##############################################################################
+##
+## Expand & Complete dataframes
+##
+##############################################################################
+
+function _expandhelper(col, vect, stride, offset)
+    for i in offset:stride:length(vect)
+        push!(col, vect[i])
+    end
+end
+
+function expand(df::AbstractDataFrame, indexcols)
+    # Check to make sure the symbols in indexcols are in the df, and check for duplicate rows in the input df
+    if !(indexcols isa Symbol || allunique(indexcols))
+        throw(ArgumentError("Elements of $indexcols must be unique"))
+    end
+    sum(nonunique(df, indexcols))>0 && @warn "duplicate rows in input; expand will only return unique combinations"
```

Comment:
Now that I look at it we could use `row_group_slots` instead of `nonunique` to squeeze out more performance (it is enough to check how many groups were created).

Reply:
Something like this?
`row_group_slots(ntuple(i -> df[!, i], ncol(df)))[1]` and comparing that to the number of rows of the input `df`?

Reply:
`row_group_slots(ntuple(i -> df[!, i], ncol(df)), Val(false))[1]` will be even more efficient for pooled column types.

Reply:
Given my later comment I think we can even avoid this - we really do not need to check for duplicates (they do no harm and if they are present `join` will expand `dummydf` as needed). If someone wants to avoid duplicates one can do it before calling `expand`.

Reply:
OK, I was just reacting to the question.

Reply:
(sure - I also wanted to learn the optimal pattern)